designs/0034-jacobian.md

```diff
@@ -0,0 +1,176 @@
+The key difference between the two is that `jacobian` will only be accumulated into for algorithms that request it. Algorithms such as maximum likelihood would request the jacobian to not be accumulated. 
+The `jacobian` keyword is available only in the `transformed parameters` block and within functions ending in `_jacobian`.
+
+
+### Example Usage
+
+The following stan program defines an `upper_bound_jacobian` function for constrainting `real` and `vector` types.
+
+```stan
+functions {
+real upper_bound_jacobian(real x, real ub) {
+  jacobian += x;
+  return upper_bound - exp(x);
+}
+vector upper_bound_jacobian(vector x, real ub) {
+  jacobian += x;
+  return upper_bound - exp(x);
+}
+}
+
+data {
+ real ub;
+ int N;
+}
+
+parameters {
+  real b_raw;
+  vector[N] b_vec_raw;
+}
+
+transformed parameters {
+  real b = upper_bound_jacobian(b_raw, ub)
+  vector[N] b = upper_bound_jacobian(b_vec_raw, ub)
+}
+```
+
+The following stan program calculates the same constraint and jacobian directly in the transformed parameters block.
+
+```stan
+data {
+ real ub;
+ int N;
+}
+
+parameters {
+  real b_raw;
+}
+
+transformed parameters {
+  jacobian += b_raw;
+  real b = upper_bound - exp(b_raw);
+}
+```
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The first thing we would do is add the `jacobian` keyword to the Stan language. 
+Like `target`, `jacobian` would only be available on the left hand side of statements to be accumulated into. 
+Its domain is restricted to the `transformed parameters` and functions ending in `_jacobian`.
+The main difference between `jacobian` and `target` is that the jacobian accumulation is optional for Stan algorithms. If an algorithm wants to do something like maximum likelihood estimation then it is possible to turn off the Jacobian accumulation.
+
+## Transformed Parameters
+
+Inside of transformed parameters we expose `jacobian` like in the functions section listed below. 
+Any jacobian accumulation will be wrapped in an `if (Jacobian)` so that the Jacobian increment only happens when requested.
+
+```stan
+parameters {
+ real b_raw;
+}
+transformed parameters {
+ // (2) Transform and accumulates like stan math directly
+ real b = exp(b_raw) + lower_bound;
+ // Underneath the hood, only actually calculated if
+ // Jacobian bool template parameter is set to true
+ jacobian += b_raw;
+}
+```
+
+## Functions ending in _jacobian
+
+Functions ending in `_jacobian` will operate the same as `_lp` functions, but instead of exposing `target` they will expose `jacobian`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Making a new keyword `jacobian` will almost surely break current user code.
+
+- Because we do not have the associated function to unconstrain the parameters users will have to initialize from parameters on the unconstrained space. 
+
+# Rationale and alternatives
```

Comment:
I think it is worth mentioning that there are alternatives (like the original `constraints` block proposal), but all the ones we can think of would build off of this, rather than being at odds with it.